core/src/word_lock.rs

```diff
@@ -46,31 +47,40 @@ impl ThreadData {
     }
 }
 
-// Returns a ThreadData structure for the current thread
-unsafe fn get_thread_data(local: &mut Option<ThreadData>) -> &ThreadData {
+// Invokes the given closure with a reference to the current thread `ThreadData`.
+fn with_thread_data<F, T>(mut f: F) -> T
+where
+    F: FnMut(&ThreadData) -> T,
+{
     // Try to read from thread-local storage, but return None if the TLS has
     // already been destroyed.
     #[cfg(has_localkey_try_with)]
-    fn try_get_tls(key: &'static LocalKey<ThreadData>) -> Option<*const ThreadData> {
-        key.try_with(|x| x as *const ThreadData).ok()
+    fn try_with_tls<F, T>(key: &'static LocalKey<ThreadData>, f: &mut F) -> Option<T>
+    where
+        F: FnMut(&ThreadData) -> T,
+    {
+        key.try_with(|x| f(x)).ok()
     }
     #[cfg(not(has_localkey_try_with))]
-    fn try_get_tls(key: &'static LocalKey<ThreadData>) -> Option<*const ThreadData> {
-        panic::catch_unwind(|| key.with(|x| x as *const ThreadData)).ok()
+    fn try_with_tls<F, T>(key: &'static LocalKey<ThreadData>, f: &mut F) -> Option<T>
+    where
+        F: FnMut(&ThreadData) -> T,
+    {
+        panic::catch_unwind(AssertUnwindSafe(|| key.with(|x| f(x)))).ok()
     }
 
     // If ThreadData is expensive to construct, then we want to use a cached
     // version in thread-local storage if possible.
     if !cfg!(windows) && !cfg!(all(feature = "nightly", target_os = "linux")) {
         thread_local!(static THREAD_DATA: ThreadData = ThreadData::new());
-        if let Some(tls) = try_get_tls(&THREAD_DATA) {
-            return &*tls;
+        if let Some(result) = try_with_tls(&THREAD_DATA, &mut f) {
```

Comment:
OK, so this is a bit subtle, but I would prefer if `f` was only called from one place and you simply extracted a reference from the thread-local storage. The reason for this is that if `f` is used in 2 different places, it will effectively be inlined twice into the function, which doubles the code size. This will require a bit of unsafe code to juggle the lifetimes.

Reply:
Ah, yeah. Makes sense. I did this now, but without much unsafe. Instead of juggling too many pointers I just stored the reference or owned value in a `Cow<ThreadData>`